autosklearn/pipeline/components/data_preprocessing/imputation/categorical_imputation.py

```diff
@@ -28,21 +29,25 @@ def fit(self, X: PIPELINE_DATA_DTYPE,
             y: Optional[PIPELINE_DATA_DTYPE] = None) -> 'CategoricalImputation':
         import sklearn.impute
 
-        fill_value = None
         if hasattr(X, 'columns'):
             kind = X[X.columns[-1]].dtype.kind
         else:
             # Series, sparse and numpy have dtype
             # Only DataFrame does not
             kind = X.dtype.kind
-        if kind in ("i", "u", "f"):
+
+        number_kinds = ("i", "u", "f")
+        if kind in number_kinds:
             # We do not want to impute a category with the default
-            # value (0 is the default) in case such default is in the
-            # train data already!
-            fill_value = 0
-            unique = np.unique(X)
-            while fill_value in unique:
-                fill_value -= 1
+            # value (0 is the default).
+            # Hence we take one greater than the max
+            unique = np.unique([*X.data, 0]) if issparse(X) else np.unique(X)
```

Comment:
`*X.data` looks like it could be a really costly operation. Could we do `np.unique(X.data)` here and then `fill_value = min(min(unique) - 1, 0)`?

Reply:
That's a better idea, I wasn't really thinking about any memory movement.

Reply:
That tests fail due to having a negative categorical label, I'll switch to max instead.

Reply:
... and switching to max seems to break other tests

Reply:
It seems like fixing this is going to take quite a while, otherwise we wait for using the `OrdinalEncoder` to take care of this. For now, I'll default it back to the original silent error that sparse matrices have `nans` encoded to 0.

Reply:
Turns out, reverting back to the orignal code now gives tests failures -_-

Reply:
This [loc](https://github.com/automl/auto-sklearn/blob/master/autosklearn/pipeline/components/data_preprocessing/imputation/categorical_imputation.py#L43) fails for sparse matrices yet I have no idea how it still works in the current master branch?

Reply:
Okay never mind it doesn't but it doesn't work as intended
```python
x = scipy.sparse.csr_matrix([[0,0,1],[0,1,0]])
unique - np.unique(x)
print(unique)

array([<2x3 sparse matrix of type '<class 'numpy.int64'>'
       	with 3 stored elements in Compressed Sparse Row format>],
      dtype=object)
      
In [10]: 0 in np.unique(x)
/home/skantify/.pyenv/versions/3.8.5/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3444: SparseEfficiencyWarning: Comparing a sparse matrix with 0 using == is inefficient, try using != instead.
  exec(code_obj, self.user_global_ns, self.user_ns)
<ipython-input-10-73ac6367503d>:1: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
  0 in np.unique(x)
  
Out[10]: False

In [11]: 1 in np.unique(x)

<ipython-input-11-053263455b76>:1: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
  1 in np.unique(x)
  
Out[11]: False
```